Enhance tools

HiranChaudhuri opened the pull request:
The editor should be a bit more usable now. The bowmen are shown on user's request.

paulwedeck commented on .gitignore:
```diff
@@ -35,3 +35,8 @@ out/
 productionStats/*.png
 productionStats/*.log
+
+# these files get easily created when running the tools from the IDE
```
You are probably running the tools from the wrong working directory. It should always be the root of the repository not jsettlers.tools/
If you do it like this then these ignores are unnecessary

HiranChaudhuri replied:
Agree. I am just hitting the 'run' button in the IDE. It works quite well, despite the fact that the editor creates these nasty files. This is just a safety so we do not add them to version control.

HiranChaudhuri replied:
Netbeans does not execute the whole project, but I can select the subprojects and hit the run button there. In that case Netbeans will execute such as

```
cd <...>/settlers-installer/app; ../gradlew --configure-on-demand /.. run
```

While it does not look wrong, the current working directory is always on the subproject.
I failed reconfiguring Netbeans to invoke this differently. But this is outside the scope of enhancing the tools only, so I will remove it for now.

paulwedeck commented on jsettlers.tools/build.gradle:
```diff
@@ -13,6 +13,10 @@ dependencies {
     implementation project(':jsettlers.network')
     implementation project(':jsettlers.testutils')
     implementation project(':jsettlers.tests')
+    // https://mvnrepository.com/artifact/org.apache.logging.log4j/log4j-core
```
This is really beyond the scope of this pr.
Also I am not going to add log4j or any other big logging framework to jsettlers (except for some very good reason ofc)

HiranChaudhuri replied:
Removed

paulwedeck commented on jsettlers.tools/src/main/java/jsettlers/buildingcreator/editor/BuildingCreatorApp.java:
```diff
@@ -117,24 +152,91 @@ private IMapInterfaceConnector startMapWindow() throws JSettlersLookAndFeelExecp
+            IBuilding.IOccupied building = (IBuilding.IOccupied)map.getBuilding();
+            occupiers = (List<IBuildingOccupier>)building.getOccupiers();
+            log.debug("OccupierPlaces:");
+            for (OccupierPlace op: variant.getOccupierPlaces()) {
```
This should be a member function of the building. You shouldn't edit the state (getOccupiers()) of the building outside of the building

paulwedeck replied:
JSettlers generally implements the following pattern:
There is one interface that describes the in game visible state of an object (like IGraphicsMovable) and there is a class that implements the game logic (like Movable).
But you can also add alternative implementation of the game logic or dummy implementations that only show a constant state (like BuildingCreatorGraphicsMovable).

PseudoBuilding does the same for buildings as BuildingCreatorGraphicsMovable does for movables.
You can just modify the logic behind its state as the needs are for the BuildingCreatorApp.

HiranChaudhuri replied:
So it would make sense to have methods like occupy() and evacuate() in IOccupied?
But how does the game today resolve the same issue? The methods do not exist today.

paulwedeck replied:
Oh no. IOccupied is just there to query what the user can see and **not** how this can change. This is purely part of the game logic and not the graphics interface.
occupy/evacuate should be functions of PseudoBuilding and not IOccupied

HiranChaudhuri replied:
Ok, I now understood what you actually pointed at. Now not IOccupied is modified but just the PseudoBuilding.

paulwedeck commented on jsettlers.tools/src/main/java/jsettlers/buildingcreator/editor/GraphicsMovable.java:
```diff
@@ -0,0 +1,127 @@
+ *
+ * @author hiran
+ */
+public class GraphicsMovable implements IGraphicsMovable {
```
I don't want to nitpick but I kind of dislike this name because it doesn't specify that this is only a dummy class and not something that should be used.
Maybe call it PseudoMovable, DummyMovable or BuildingCreatorMovable?
I think FakeMovable is taken

paulwedeck commented:
I cleaned the code up a bit and I also fixed the swordsmen.
Please ident your files with tabs instead of spaces.
From what I can see its not yet possible to update the positions or save them (they aren't in the show xml dialog).
Also maybe consider to to put the tools in a ButtonGroup as radio buttons and add the occupier stuff as its own "tool". This way you could hide the markings.

HiranChaudhuri commented:
I am not yet sure about indentation - reconfigured Netbeans to not replace tabs but I also do not check for whitespace manually.
Positions can be added, updated and removed, and the result is visible in the xml dialog.

paulwedeck commented on jsettlers.tools/src/main/java/jsettlers/buildingcreator/editor/BuildingCreatorApp.java:
```diff
@@ -68,81 +99,195 @@ public class BuildingCreatorApp implements IMapInterfaceListener, Runnable {
+            switch (definition.getBuilding().getType()) {
+                case BIG_TOWER:
+                case CASTLE:
+                case LOOKOUT_TOWER:
```
lookout towers can't have soldiers inside. Generally speaking every building can't have occupierplaces so I don't see a strong reason to limit it to military buildings

HiranChaudhuri replied:
"every building can't have occupierplaces" probably means that potentially every building can have OccupierPlaces? If that is the case, I will just drop that limitation.
